File: src/io/github/hulang1024/chinesechess/rule/AbstractChess.java
package io.github.hulang1024.chinesechess.rule;

/**
 * 抽象的棋子
 * @author Hu Lang
 */
public abstract class AbstractChess {
    /** 当前位置 */
    public ChessPosition pos;
    /** 所属军方 */
    public HostEnum host;

    /**
     * 判断指定位置是否可走。
     * 不考虑指定位置是否为空或有棋子，不应包含全局规则。
     * @param destPos
     * @param game
     * @return
     */
    public abstract boolean canGoTo(ChessPosition destPos, ChineseChessGame game);
}

File: client/src/rulesets/chinesechess/chess/AbstractChess.ts
import Chess from "../Chess";
import ChessPos from "../ChessPos";
import ChessHost from "../../chess_host";
import Game from "../Game";

/**
 * 抽象的棋子
 */
export default abstract class AbstractChess implements Chess {
  /** 当前位置 */
  protected pos: ChessPos;

  /** 所属棋方 */
  protected host: ChessHost;

  /** 是否是正面 */
  protected front = true;

  constructor(pos: ChessPos, host: ChessHost) {
    this.pos = pos;
    this.host = host;
  }

  abstract canGoTo(destPos: ChessPos, game: Game): boolean;

  setPos(pos: ChessPos) {
    this.pos = pos;
  }

  getPos() {
    return this.pos;
  }

  getHost() {
    return this.host;
  }

  setFront(b: boolean) {
    this.front = b;
  }

  isFront() {
    return this.front;
  }

  is(chessClass: any) {
    return this instanceof chessClass;
  }
}
